camel/storages/graph_storages/neo4j_graph.py

```diff
@@ -0,0 +1,547 @@
+        except (
+                ClientError
+        ):  # Read-only user might not have access to schema information
+            constraint = []
+            index = []
+
+        self.structured_schema = {
+            "node_props":
+            {el["labels"]: el["properties"]
+             for el in node_properties},
+            "rel_props":
+            {el["type"]: el["properties"]
+             for el in rel_properties},
+            "relationships": relationships,
+            "metadata": {
+                "constraint": constraint,
+                "index": index
+            }
+        }
+
+        # Format node properties
+        formatted_node_props = []
+        for el in node_properties:
+            props_str = ", ".join([
+                f"{prop['property']}: {prop['type']}"
+                for prop in el["properties"]
+            ])
+            formatted_node_props.append(f"{el['labels']} {{{props_str}}}")
+
+        # Format relationship properties
+        formatted_rel_props = []
+        for el in rel_properties:
+            props_str = ", ".join([
+                f"{prop['property']}: {prop['type']}"
+                for prop in el["properties"]
+            ])
+            formatted_rel_props.append(f"{el['type']} {{{props_str}}}")
+
+        # Format relationships
+        formatted_rels = [
+            f"(:{el['start']})-[:{el['type']}]->(:{el['end']})"
+            for el in relationships
+        ]
+
+        self.schema = "\n".join([
+            "Node properties are the following:",
+            ", ".join(formatted_node_props),
+            "Relationship properties are the following:",
+            ", ".join(formatted_rel_props),
+            "The relationships are the following:",
+            ", ".join(formatted_rels),
+        ])
+
+    def add_triplet(self, subj: str, obj: str, rel: str) -> None:
+        r"""Adds a relationship (triplet) between two entities in the database.
+
+        Args:
+            subj (str): The identifier for the subject entity.
+            obj (str): The identifier for the object entity.
+            rel (str): The relationship between the subject and object.
+        """
+        query = """
+            MERGE (n1:`%s` {id:$subj})
+            MERGE (n2:`%s` {id:$obj})
+            MERGE (n1)-[:`%s`]->(n2)
+        """
+
+        prepared_statement = query % (
+            BASE_ENTITY_LABEL.replace("_", ""),
+            BASE_ENTITY_LABEL.replace("_", ""),
+            rel.replace(" ", "_").upper(),
+        )
+
+        # Execute the query within a database session
+        with self.driver.session(database=self.database) as session:
+            session.run(prepared_statement, {"subj": subj, "obj": obj})
+
+    def _delete_rel(self, subj: str, obj: str, rel: str) -> None:
+        r"""Deletes a specific relationship between two nodes in the Neo4j
+        database.
+
+        Args:
+            subj (str): The identifier for the subject entity.
```

Comment:
Does subj here means source node? Maybe `source` and `target` are a little bit better?

Reply:
`subject ` is widely used in knowledge graph triplet

Reply:
Then maybe also change the `Relationship` attribute from source to subject etc.?

Reply:
updated